apps/condo/domains/billing/schema/fields/BillingReceipt/Services.js

```diff
@@ -0,0 +1,96 @@
+const Ajv = require('ajv')
+const { ToPayDetailsFields } = require('@condo/domains/billing/schema/fields/BillingReceipt/ToPayDetailsField')
+const {
+    BILLING_RECEIPT_SERVICE_TO_PAY_DETAILS_FIELD_NAME,
+    BILLING_RECEIPT_SERVICE_TO_PAY_DETAILS_INPUT_NAME,
+    BILLING_RECEIPT_SERVICE_FIELD_NAME,
+    BILLING_RECEIPT_SERVICE_INPUT_NAME,
+} = require('@condo/domains/billing/constants')
+const { render, getValidator } = require('@condo/domains/billing/schema/fields/utils/json.utils')
+const { Json } = require('@core/keystone/fields')
+
+const ServiceToPayDetailsFields = {
+    ...ToPayDetailsFields,
+    volume: 'String',
+    tariff: 'String',
+    measure: 'String',
+}
+
+const ServiceFields = {
+    id: 'String',
+    name: 'String!',
+    toPay: 'String!',
+    toPayDetails: BILLING_RECEIPT_SERVICE_TO_PAY_DETAILS_FIELD_NAME,
+}
+const ServiceInputs = {
+    ...ServiceFields,
+    toPayDetails: BILLING_RECEIPT_SERVICE_TO_PAY_DETAILS_INPUT_NAME,
+}
+
+
+const SERVICES_GRAPHQL_TYPES = `
+    type ${BILLING_RECEIPT_SERVICE_TO_PAY_DETAILS_FIELD_NAME} {
+        ${render(ServiceToPayDetailsFields)}
+    }
+    
+    input ${BILLING_RECEIPT_SERVICE_TO_PAY_DETAILS_INPUT_NAME} {
+        ${render(ServiceToPayDetailsFields)}
+    }
+
+    type ${BILLING_RECEIPT_SERVICE_FIELD_NAME} {
+        ${render(ServiceFields)}
+    }
+    
+    input ${BILLING_RECEIPT_SERVICE_INPUT_NAME} {
+        ${render(ServiceInputs)}
+    }
+`
+
+const ServiceToPayDetailsSchema = {
+    type: ['object', 'null'],
+    properties: Object.assign({},
+        ...Object.keys(ServiceToPayDetailsFields).map((field) => ({ [field]: { type: ['string', 'null'] } })),
+        { formula: { type: 'string' } },
+    ),
+}
+
+const ServicesSchema = {
+    type: 'array',
+    items: {
+        type: 'object',
+        properties: {
+            id: { type: ['string', 'null'] },
+            name: { type: 'string' },
+            toPay: { type: 'string' },
+            toPayDetails: ServiceToPayDetailsSchema,
+        },
+        // todo(toplenboren) discuss the analytics and standartization service for services
+        required: ['name', 'toPay'],
+        additionalProperties: false,
+    },
+}
+
+const ajv = new Ajv()
+const ServicesSchemaValidator = ajv.compile(ServicesSchema)
+
+const validateServices = getValidator(ServicesSchemaValidator)
+const SERVICE_TO_PAY_DETAILS_QUERY_LIST = Object.keys(ServiceToPayDetailsFields).join(' ')
+const SERVICES_QUERY_LIST = `${Object.keys(ServiceFields).join(' ')} { ${SERVICE_TO_PAY_DETAILS_QUERY_LIST} }`
+
+const SERVICES_FIELD = {
+    schemaDoc: 'Services to pay for. Every service has id, name and toPay. Service may or may not have toPay detail. Detail level 3 and 4',
+    type: Json,
+    isRequired: false,
+    extendGraphQLTypes: [SERVICES_GRAPHQL_TYPES],
+    graphQLInputType: `[${BILLING_RECEIPT_SERVICE_INPUT_NAME}!]`,
```

Comment:
it looks like a substitution of definitions. The name `BillingReceiptServiceFieldInput` assume to be an input of the field. But, the real input type is `[BillingReceiptServiceFieldInput!]`

Reply:
It's confusing me

Reply:
Why? Field called services, type is service….So services =[service]…seems like intuitive

Reply:
The problem here is that variable

`BILLING_RECEIPT_SERVICE_INPUT_NAME` is simply `BillingReceiptServiceFieldInput`

But if real world it should be used like this: `[ + BILLING_RECEIPT_SERVICE_INPUT_NAME + ]` which is super counter-intuitive

Imagine if we need to use this variable in other context or other file, in that case we should somehow know that we need to surround it with `[ ]`

Maybe we need to make `BILLING_RECEIPT_SERVICE_FIELD_NAME` equal to real value of the field `[BillingReceiptServiceFieldInput!]`

Reply:
Moved to const 👍